overlord/snapstate/handlers.go

```diff
@@ -1056,6 +1118,20 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 		return err
 	}
 
+	// Restore disabled services
+	var missingSvcs []string
+	missingSvcs, err = m.backend.RestoreDisabledServices(newInfo, snapst.LastActiveDisabledServices, pb)
+	if err != nil {
+		// nothing to undo, since worst case we would have just disabled some
```

Comment:
Probably part of more general issue with task handlers. When a do handler fails, the undo handler is not called for that task. Undo is only called for tasks that reached DoneState. Instead, the do handler is expected to perform a cleanup of all work that it did up to the point where the error occurred. In this case, I believe, we should have a call to UnlikSnap() here. 

Best check with @mvo or @pedronis though, I always get confused by this.

Reply:
Ah thanks for clarifying this I was wondering about that. In this case I will also add some unit tests around RestoreDisabledServices failing and triggering an unlink as well here

Reply:
ah thanks for clarifying this, it wasn't super clear to me reading the code. In that case, I agree with you and we should do an Unlink here as well.

Reply:
hmm, so looking at this, we actually have a number of other places where if we encounter an error, Unlink is not called but probably should be, no? for example:

* https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1062
* https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1068
* https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1093
* https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1108
* https://github.com/snapcore/snapd/blob/master/overlord/snapstate/handlers.go#L1120

Maybe we should refactor this code to match the below so that we catch all of these sorts of "forgot to cleanup something" type bugs in this code:

https://github.com/snapcore/snapd/blob/583fe122f1e8ea5ad99964d322dab02da6e7614b/wrappers/services.go#L210-L228

Reply:
Well I ended up implementing a defer'd unlink cleanup function, if this is wrong for some reason I will refactor again, but in the meantime I did add a unit test to ensure that if RestoreDisabledServices does fail, we do an unlink to cleanup.